### PR TITLE
[webview_flutter] Support basic auth requests

### PR DESCRIPTION
**Pull request by wwwdata**
## Description

This enables the webview to visit http basic auth protected websites.

## Related Issues

Originally, it was attempted to implement this here: https://github.com/flutter/plugins/pull/1283
But the author then decided to split up the features into smaller pull requests. However, the one for auth requests was never opened, only for the other features: https://github.com/flutter/plugins/pull/1389

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by FuturistLabs** [comment]
Any idea when this feature will be available?
Thanks in advance!

**Comment by wwwdata** [comment]
> Any idea when this feature will be available?
> Thanks in advance!

As soon as someone from Google presses the green merge button here... Which unfortunately takes a very long time.

**Comment by JJibreel** [comment]
@wwwdata thank you for working on this feature. It seems that its nearing completion, cant wait!

**Comment by stuartmorgan** [comment]
We apologize for the long delay in triaging this PR. We’re in the process of overhauling our PR triage system to respond much more quickly, as well as working through the backlog.

It looks like this should be pretty straightforward to review, and has good test coverage. If you're still interested in moving forward with this, once you've updated this to apply to the current version of the plugin I can review it.

**Comment by wwwdata** [comment]
Hi, I will rebase this then and make it up to date when I have time.

**Comment by wwwdata** [comment]
@stuartmorgan ok I just rebased it and fixed all the conflicts.

File: packages/webview_flutter/lib/platform_interface.dart
Comment (stuartmorgan) [inline packages/webview_flutter/lib/platform_interface.dart:41]: This class is public API, so adding a new method without an implementation is a breaking change unless I'm missing something. This should probably have a default implementation that returns null.

File: packages/webview_flutter/lib/src/webview_method_channel.dart
Comment (stuartmorgan) [inline packages/webview_flutter/lib/src/webview_method_channel.dart:29]: `Object?`

File: packages/webview_flutter/lib/webview_flutter.dart
Comment (stuartmorgan) [inline packages/webview_flutter/lib/webview_flutter.dart:855]: As a top-level public class, this should probably have a less collision-prone name, like `WebViewAuthInfo`.

File: packages/webview_flutter/lib/webview_flutter.dart
Comment (stuartmorgan) [inline packages/webview_flutter/lib/webview_flutter.dart:857]: const

Comment (stuartmorgan) [inline packages/webview_flutter/lib/webview_flutter.dart:857]: Since these have the same type, it would be safer to use `required` named arguments rather than positional arguments.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
Small comments below, but more importantly: now that I'm reviewing this in detail I see that the test coverage isn't actually complete; it only has Dart unit tests, so none of the native code is being exercised. This should have an integration test if at all possible (and failing that, native unit tests to complement the Dart unit test).

**Comment by stuartmorgan** [comment]
(FYI the mass of CI failures is from an infrastructure issue we hit; the next time you rebase it'll clear up.)

File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebViewClient.java
Comment (wwwdata) [reply packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebViewClient.java:175]: I don't know. How would I report that? This is basically handling the response from dart where it should definitely always be the case that `username` and `password` were sent. I am just checking it because the types are always optional when you send them over the MethodChannel.

File: packages/webview_flutter/lib/platform_interface.dart
Comment (wwwdata) [reply packages/webview_flutter/lib/platform_interface.dart:41]: There is only one implementation in the `_PlatformCallbacksHandler` of the webview flutter widget. Where the new method is implemented. So it's not breaking anything, or?

File: packages/webview_flutter/lib/webview_flutter.dart
Comment (hpoul) [inline packages/webview_flutter/lib/webview_flutter.dart:155]: Is there any reason this has to be synchronous? Wouldn't it make more sense to have this `async` (ie. returning `Future<WebViewAuthInfo>`) I could imagine that retrieving credentials (or asking the user) will not always be possible synchronously.

Comment (wwwdata) [reply packages/webview_flutter/lib/webview_flutter.dart:155]: Yes, that's what I also thought. I will change it to be async.

Comment (hpoul) [reply packages/webview_flutter/lib/webview_flutter.dart:155]: 👍️ just discovered that it should probably be `Future<WebViewAuthInfo?>` return type (ie. nullable), right now the callback has no way of indicating that it has no auth info available?

File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebViewClient.java
Comment (stuartmorgan) [reply packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebViewClient.java:175]: If the Dart side cannot send null, you can just assert that they aren't null rather than checking.

File: packages/webview_flutter/lib/platform_interface.dart
Comment (stuartmorgan) [reply packages/webview_flutter/lib/platform_interface.dart]: > There is only one implementation

If an interface is part of a public API surface, then you have no way of knowing how many implementations there are.

Comment (wwwdata) [reply packages/webview_flutter/lib/platform_interface.dart]: ok added a null default

**Comment by wwwdata** [comment]
just pushed my changes again. I hope it's ok now. Sorry for the delay I was on vacation

**Comment by hpoul** [comment]
@stuartmorgan is there still something missing in this PR? 🤔️ (except fixing the changelog conflict) thanks.

**Review by stuartmorgan-g — COMMENTED** [review]
I'm sorry this review fell through the cracks again 🙁 

Overall this looks good; the complication now is that the plugin was converted to fully federated mode in the intervening time so this will now need to be [split into multiple PRs](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-federated-plugins). There's also work underway at the moment to change the communication channels, which would create a second round of conflicts.

/cc @bparrishMines How would you prefer to move forward here relative to your channel work? I think the options are:
1) we split the PR ASAP, get it landed, and then rebase your work with the changes, or
2) you finish your method channel changes, then we update this PR to adapt to both sets of changes at once.

Either way @wwwdata, I can take care of the updating and splitting of this PR if you'd prefer, since all the added work here is due to my losing track of the review.

**Comment by stuartmorgan** [comment]
Per discussion with @bparrishMines I'm going to hold off on updating this until the in-progress conversion is complete.

**Comment by hpoul** [comment]
Is there still a conversion in progress? I guess the original federated plugin changes are done? @stuartmorgan  I have seen the iOS implementation has still some refactoring in progress? https://github.com/flutter/plugins/pull/4626
I guess working on this still doesn't make sense while that is still going on? 🤔️

**Comment by stuartmorgan** [comment]
Triage update: iOS webview conversion is still in progress, so this still blocked for the moment.

**Comment by Hixie** [comment]
Latest status is that we're still waiting on https://github.com/flutter/flutter/issues/93732.
Before this can land even after that is fixed, the conflicts will need to be resolved.

**Comment by stuartmorgan** [comment]
The Dart conversion is complete, but in the meantime we found https://github.com/flutter/flutter/issues/94051 which blocks this, since we currently implement WebViewPlatformCallbacksHandler in our implementations, making changes to it breaking changes. We're actively working on that issue, and hope to have it resolved soon.

**Comment by stuartmorgan** [comment]
Status update from triage: this is still blocked, but we hope to have it unblocked soon.

**Comment by stuartmorgan** [comment]
This is unblocked now that v4 has landed. @bparrishMines This is one we should look at updating.

**Comment by stuartmorgan** [comment]
Update from triage: this is still on our list to update; we're currently working through updating a couple of higher-priority webview features, but we'll get to this soon.

**Comment by stuartmorgan-g** [comment]
We've just completed the migration of the plugin code to the flutter/packages repository, as described in https://flutter.dev/go/flutter-plugins-repo-migration, and this repository is now being archived. Unfortunately that means that all in-progress PRs here must be moved to flutter/packages.

Per discussion above about this PR having fallen through the cracks, either I or @bparrishMines will be opening an updated version of this PR in flutter/packages, so there's no need for any action here. Per my comment above, we expect to get to this fairly soon.

**Pull request closed without merging**
